R/XcmsExperiment-functions.R

```diff
@@ -1016,3 +1030,71 @@ featureArea <- function(object, mzmin = min, mzmax = max, rtmin = min,
         .require_spectra()
         Spectra::Spectra(res)
     }
+
+#' helper function to get indices of chromatographic peaks for eventual
+#' subsetting with `rt`, `mz`, `msLevel` and `type`.
+#'
+#' @param object `XcmsExperiment` or `XCMSnExp` object with `chromPeaks`.
+#'
+#' @return `integer` vector with the indices of the `chromPeaks` matching the
+#'     requested filtering.
+#'
+#' @noRd
+.index_chrom_peaks <- function(object, rt = numeric(),
+                               mz = numeric(), ppm = 0,
+                               msLevel = integer(),
+                               type = c("any", "within",
+                                        "apex_within")) {
+    type <- match.arg(type)
+    pks <- .chromPeaks(object)
+    keep <- rep(TRUE, nrow(pks))
+    if (length(msLevel))
+        keep <- keep &
+            chromPeakData(
+                object, return.type = "data.frame")$ms_level %in% msLevel
+    ## Select peaks within rt range.
+    if (length(rt)) {
+        rt <- range(as.numeric(rt))
+        if (type == "any")
+            keep <- keep & pks[, "rtmin"] <= rt[2L] & pks[, "rtmax"] >= rt[1L]
+        if (type == "within")
+            keep <- keep & pks[, "rtmin"] >= rt[1L] & pks[, "rtmax"] <= rt[2L]
+        if (type == "apex_within")
+            keep <- keep & pks[, "rt"] >= rt[1L] & pks[, "rt"] <= rt[2L]
+    }
+    ## Select peaks within mz range, considering also ppm
+    if (length(mz)) {
+        mz <- range(as.numeric(mz))
+        if (is.finite(mz[1L]))
+            mz[1L] <- mz[1L] - mz[1L] * ppm / 1e6
+        if (is.finite(mz[2L]))
+            mz[2L] <- mz[2L] + mz[2L] * ppm / 1e6
+        if (type == "any")
+            keep <- keep & pks[, "mzmin"] <= mz[2L] & pks[, "mzmax"] >= mz[1L]
+        if (type == "within")
+            keep <- keep & pks[, "mzmin"] >= mz[1L] & pks[, "mzmax"] <= mz[2L]
+        if (type == "apex_within")
+            keep <- keep & pks[, "mz"] >= mz[1L] & pks[, "mz"] <= mz[2L]
+    }
+    which(keep)
+}
+
+#' Helper function to return the chromPeakData as-is (as a data.frame) from
+#' a `XcmsExperiment` object.
+#'
+#' @noRd
+.chromPeakData <- function(object, msLevel = integer()) {
+    if (length(msLevel))
+        object@chromPeakData[object@chromPeakData$ms_level %in% msLevel, ]
+    else object@chromPeakData
+}
+
+#' Helper function to return the **full**  chromPeaks matrix without any
+#' filtering from either a `XcmsExperiment` or `XCMSnExp` object
+#'
+#' @noRd
+.chromPeaks <- function(object) {
```

Comment:
That function was required, because we would end up in (infinitely) recursive call with the `chromPeaks` method: `chromPeaks` uses `.index_chrom_peaks`, which in turn calls `chromPeaks` (that again calls `.index_chrom_peaks`, ...). To extract the **full** chrom peaks matrix we can use `.chromPeaks`, to get eventually filtered chrom peaks we use `chromPeaks` (that's what the user will use).